### PR TITLE
Moving Aggregation to Java

### PR DESCRIPTION
**Pull request by jdunkerley**
### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso [hunk 1]
```diff
@@ -7,6 +7,7 @@ polyglot java import org.enso.table.data.table.Column as Java_Column
 polyglot java import org.enso.table.operations.OrderBuilder
 
 ## Creates a new column given a name and a vector of elements.
+   Testing 123 Who's Here
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Column.enso]: :)

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Aggregator.enso [hunk 2]
```diff
@@ -148,3 +150,21 @@ new table column =
                 v = col.at i
                 if v.is_nothing then m else if m.is_nothing then v else if m.length >= v.length then m else v
             create_aggregator accumulator=accumulator
+
+java_aggregator : Table->Aggregate_Column->AggregateColumnDefinition
+java_aggregator table name column =
+    resolve c = column.resolve_column table c
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Aggregator.enso]: I would suggest switching to the pattern where we do the resolution first (probably as we discussed simply as part of validation) and then in other helper functions just assuming that the columns are properly resolved. Thus we can avoid this `resolve c` pattern being sprinkled everywhere.

Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Aggregator.enso]: (But I guess it can only be done after the validation with resolution is merged so I understand it may not be possible yet :) )

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso [hunk 3]
```diff
@@ -486,6 +486,15 @@ type Table
...
 
+    aggregate_2 : [Aggregate_Column] -> Table
+    aggregate_2 columns =
+        validated = Aggregate_Column_Helper.validate columns this
+
+        java_columns = validated.key_columns.map .java_column
+        index = this.java_table.indexFromColumns java_columns.to_array
+
+        new_table = validated.valid_columns.map c->(Aggregate_Column_Aggregator.java_aggregator this c.first c.second)
+        index.makeTable new_table.to_array
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso]: This seems to return a raw Java table, so I think you should wrap it in the Enso wrapper.

File: std-bits/table/src/main/java/org/enso/table/data/index/MultiValueIndex.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+  public MultiValueIndex(Storage[] keyColumns) {
+    this.locs = new HashMap<>();
+    if (keyColumns.length != 0) {
+      int size = keyColumns[0].size();
+      for (int i = 0; i < size; i++) {
+        int finalI = i;
+        MultiValueKey key = new MultiValueKey(Arrays.stream(keyColumns).map(c -> c.getItemBoxed(finalI)).toArray());
+        List<Integer> its = this.locs.computeIfAbsent(key, x -> new ArrayList<>());
+        its.add(i);
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/index/MultiValueIndex.java]:
```suggestion
        List<Integer> ids = this.locs.computeIfAbsent(key, x -> new ArrayList<>());
        ids.add(i);
```

Not sure what `its` would stand for here?

**Comment by jdunkerley** [comment]
| Test | Ref | New |
| --- | --- | --- |
| Count table | 122.4 | 6.1 |
| Max table | 166.3 | 5.0 |
| Sum table | 120.6 | 4.0 |
| StDev table | 50.2 | 2.6 |
| Count grouped | 191.4 | 5.5 |
| Max table | 166.3 | 5.0 |
| Sum table | 120.6 | 4.0 |
| StDev grouped | 118.9 | 4.0 |
| Count 2 level groups | 468.5 | 3.7 |
| Max table | 166.3 | 5.0 |
| Sum table | 120.6 | 4.0 |
| StDev 2 level groups | 254.8 | 5.1 |

File: app/ide-desktop/package-lock.json [hunk 5]
```diff
@@ -8,34 +8,29 @@
 			"name": "root",
 			"version": "1.0.0",
 			"dependencies": {
-				"axios": "^0.21.0",
```
Comment (radeusgd) [inline app/ide-desktop/package-lock.json]: What is the reason for this change? You are not really modifying rust related stuff, so it is surprising.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso [hunk 6]
```diff
@@ -142,3 +144,32 @@ resolve_aggregate table problem_builder aggregate_column =
...
+        Percentile p c _ -> AggregateColumnDefinition.Percentile name c.java_column p
+        Mode c _ -> AggregateColumnDefinition.Mode name c.java_column
+        First c _ ignore_nothing _ -> AggregateColumnDefinition.First name c.java_column ignore_nothing
+        Last c _ ignore_nothing _ -> AggregateColumnDefinition.Last name c.java_column ignore_nothing
+        Maximum c _ -> AggregateColumnDefinition.MinOrMax name c.java_column 1
+        Minimum c _ -> AggregateColumnDefinition.MinOrMax name c.java_column -1
+        Shortest c _ -> AggregateColumnDefinition.MinOrMaxLength name c.java_column -1
+        Longest c _ -> AggregateColumnDefinition.MinOrMaxLength name c.java_column 1
+        Concatenate c _ join prefix suffix quote -> AggregateColumnDefinition.Concatenate name c.java_column join prefix suffix quote
+        _ -> Error.throw "Unsupported aggregation."
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso]: I think we shouldn't throw pure `Text` values. This will render badly in the GUI.

I know the codebase still has some places like this, but whenever we modify something we should go in the direction of avoiding this. I think `Illegal_Argument_Error` or `Illegal_State_Error` are the universal wrappers to use, or in this case maybe it's worth creating a separate class? Like the `Unsupported_Database_Operation_Error` we were talking about.

File: std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  protected static Long CastToLong(Object value) {
+    if (value instanceof Long) {
+      return (Long)value;
+    } else if (value instanceof Integer) {
+      return ((Integer)value).longValue();
+    } else if (value instanceof Byte) {
+      return ((Byte)value).longValue();
+    } else if (value instanceof Float && ((Float)value) % 1 == 0) {
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: What's the purpose of the `((Float)value) % 1 == 0` check? Is it checking if it is an integer value? I'd suggest adding some temporary variable to give this a name to make the intent clearer.

File: std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+      return ((Float)value).longValue();
+    } else if (value instanceof Double && ((Double)value) % 1 == 0) {
+      return ((Double)value).longValue();
+    }
+
+    return null;
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: I think it may be worth to check here for `BigInteger` and throw some exception saying that currently this is not supported.

A regular big Enso number magically becoming a `null` seems like a scenario for data corruption.

File: std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    if (current == null) {
+      return null;
+    }
+
+    if (prefix != null) { current.insert(0, prefix); }
+    current.append(suffix);
+    return current.toString();
+  }
+
+  private static String ToQuotedString(Object value, final String quote, final String join) {
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java]: nitpick: but I think we start with `lowerCase` for functions, even static ones, don';t we?

File: std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+  public AggregatedProblems getProblems() {
+    return problems;
+  }
+
+  /***
+   * Compute the value for a set of rows
+   * @param rows - indexes to the rows in the source table to aggregate on
+   * @return aggregated value
+   */
+  public Object aggregate(int[] rows) {
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]:
```suggestion
  public Object aggregate(long[] rows) {
```

I'm pretty sure our numbers are `long` by default, I don't remember seeing `int` in such contexts.

Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: Also if we have a specialization for integer columns to avoid boxing, may be worth to also have one for `double` columns?

File: std-bits/table/src/main/java/org/enso/table/data/table/problems/AggregatedProblems.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
+package org.enso.table.data.table.problems;
+
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+public class AggregatedProblems {
+  private final Problem[] problems;
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/table/problems/AggregatedProblems.java]: Why not use some ArrayList?

File: std-bits/table/src/main/java/org/enso/table/data/table/aggregate/AggregateColumnDefinition.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
+package org.enso.table.data.table.aggregate;
+
+import org.enso.table.aggregations.*;
+import org.enso.table.data.table.Column;
+
+public class AggregateColumnDefinition {
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/data/table/aggregate/AggregateColumnDefinition.java]: I don't think these aliases are really necessary, we can call the constructors from Enso by `ClassName.new`.

Unless you think it is worth keeping them to have all builders in one place - I guess it is a good enough reason.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso [hunk 13]
```diff
@@ -142,3 +144,32 @@ resolve_aggregate table problem_builder aggregate_column =
...
+        Count_Not_Empty c _ -> AggregateColumnDefinition.CountEmpty name c.java_column False
+        Count_Empty c _ -> AggregateColumnDefinition.CountEmpty name c.java_column True
+        Sum c _ -> AggregateColumnDefinition.Sum name c.java_column
+        Average c _ -> AggregateColumnDefinition.Mean name c.java_column
+        Standard_Deviation c _ p -> AggregateColumnDefinition.StDev name c.java_column p
+        Median c _ -> AggregateColumnDefinition.Percentile name c.java_column 0.5
+        Percentile p c _ -> AggregateColumnDefinition.Percentile name c.java_column p
+        Mode c _ -> AggregateColumnDefinition.Mode name c.java_column
+        First c _ ignore_nothing _ -> AggregateColumnDefinition.First name c.java_column ignore_nothing
+        Last c _ ignore_nothing _ -> AggregateColumnDefinition.Last name c.java_column ignore_nothing
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso]: What about `order_by`?

**Review by radeusgd — COMMENTED** [review]
The shape looks good in general, but have a few comments and I still need to have a look as I didn't manage to review everything.

One thing I'm a bit worried about is if we are not doing too much boxing for integer/double columns. This can introduce some performance cost.

We have the "old" aggregates in our Table using the `group by=...` function which I think are using streams to potentially avoid boxing. There are only most basic operations implemented there (sum, min, max, mean), but it may be good to create a comparative benchmark comparing the performance of old and new aggregations to see where we are at.

I'm not sure if this should be done as part of this particular PR, but I think this is something we should check at some point.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Aggregator.enso (hunk 2 above)
Comment (radeusgd) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Aggregator.enso]: I forgot to submit this and had this queued but I think its irrelevant now.

File: std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java (hunk 10 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: these are the indexes - a bad name of the variable.

File: std-bits/table/src/main/java/org/enso/table/data/table/aggregate/AggregateColumnDefinition.java (hunk 12 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/data/table/aggregate/AggregateColumnDefinition.java]: Left over from the previous iteration will remove

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso (hunk 13 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso]: will do in another ticket

File: app/ide-desktop/package-lock.json (hunk 5 above)
Comment (jdunkerley) [reply app/ide-desktop/package-lock.json]: Reverted - happened when I ran the IDE build on mine

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso [hunk 14]
```diff
@@ -492,44 +491,18 @@ type Table
...
-                new = accumulator current i
-                array . set_at row_index new
-
-            map.insert key row_index
-
         on_problems.attach_problems_before validated.problems <|
-            # Build Table
-            if ((validated.key_columns.length == 0) && (this.row_count == 0)) then (add_row Nothing) else
-                0.up_to this.row_count . fold Map.empty aggregate
+            java_columns = validated.key_columns.map .java_column
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso]: Would be good if the name made it clear that these are key columns. Maybe `java_key_columns`?

Just `java_columns` sounds like any set of columns - all columns from this table? Or something. I know it is a bit of a nitpick, but I think this helps readability quite a lot.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso [hunk 15]
```diff
@@ -67,3 +66,31 @@ type Input_Indices_Already_Matched (indices : [Integer])
...
     "The indices "+this.indices.short_display_text+" matched columns which have been matched earlier by other indices, so they did not introduce any new columns into the result."
+
+## Indicates that an aggregation calculation could not be completed.
+type Invalid_Aggregation (column:Text) (row:Integer) (message:Text)
+
+Invalid_Aggregation.to_display_text : Text
+Invalid_Aggregation.to_display_text =
+    "The "+this.column+" could not be calculated at "+this.row.to_text+" : "+this.message
+
+## Indicates that an floating point number was used in a grouping.
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso]:
```suggestion
## Indicates that a floating point number was used in a grouping.
```

File: distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso [hunk 16]
```diff
@@ -67,3 +66,31 @@ type Input_Indices_Already_Matched (indices : [Integer])
...
+Floating_Point_Grouping.to_display_text : Text
+Floating_Point_Grouping.to_display_text =
+    "Grouping on floating points is not recommended within "+this.column+" at row "+this.row.to_text+"."
+
+## Indicates that a text value with a delimiter was included in a concatenation without any quote character
+type Unquoted_Delimiter (column:Text) (row:Integer)
+
+Unquoted_Delimiter.to_display_text : Text
+Unquoted_Delimiter.to_display_text =
+    "The "+this.column+" at row "+this.row.to_text+" contained an unquoted delimiter."
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso]: I think some further explanation would be helpful. Because just looking at this, if I didn't discuss with you the behaviour of `Concat` in more detail, I wouldn't really know what is the issue here.

Wondering if it would make sense to have some optional `details : Text` field on problems which the GUI could show as a tooltip/foldable message which could include some practical insights into each warning. Not saying this should be done now, but I think it would be very helpful for the users at some point. But probably something to consider at a later polishing stage.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso [hunk 17]
```diff
@@ -67,3 +66,31 @@ type Input_Indices_Already_Matched (indices : [Integer])
...
+
+## Warning when additional warnings occurred.
+type Additional_Warnings (count:Integer)
+
+Additional_Warnings.to_display_text : Text
+Additional_Warnings.to_display_text =
+    "There were additionally "+this.count.to_text+" additional issues."
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso]:
```suggestion
Additional_Warnings.to_display_text : Text
Additional_Warnings.to_display_text =
    "There were "+this.count.to_text+" additional issues."
```

`additionally N additional issues` sounds very repetitive, I'd rephrase this somehow - one suggestion above.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso [hunk 18]
```diff
@@ -8,7 +8,27 @@ import Standard.Table.Internal.Problem_Builder
 import Standard.Table.Internal.Unique_Name_Strategy
 import Standard.Table.Internal.Table_Helpers
 
-from Standard.Table.Error as Error_Module import Missing_Input_Columns, Column_Indexes_Out_Of_Range, No_Output_Columns, Duplicate_Output_Column_Names, Invalid_Output_Column_Names
+from Standard.Table.Error as Error_Module import Missing_Input_Columns, Column_Indexes_Out_Of_Range, No_Output_Columns, Duplicate_Output_Column_Names, Invalid_Output_Column_Names, Invalid_Aggregation, Floating_Point_Grouping, Unquoted_Delimiter, Additional_Warnings
+
+polyglot java import org.enso.table.aggregations.Concatenate as ConcatenateColumn
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso]: I'm wondering if `ConcatenateColumn` is a good name. Maybe `ConcatenateAggregation`? It is not really a column but a column builder which performs the aggregation, right?

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso [hunk 19]
```diff
@@ -142,3 +162,48 @@ resolve_aggregate table problem_builder aggregate_column =
...
+        Count_Not_Empty c _ -> CountEmptyColumn.new name c.java_column False
+        Count_Empty c _ -> CountEmptyColumn.new name c.java_column True
+        Sum c _ -> SumColumn.new name c.java_column
+        Average c _ -> MeanColumn.new name c.java_column
+        Standard_Deviation c _ p -> StandardDeviationColumn.new name c.java_column p
+        Median c _ -> PercentileColumn.new name c.java_column 0.5
+        Percentile p c _ -> PercentileColumn.new name c.java_column p
+        Mode c _ -> ModeColumn.new name c.java_column
+        First c _ ignore_nothing _ -> FirstColumn.new name c.java_column ignore_nothing
+        Last c _ ignore_nothing _ -> LastColumn.new name c.java_column ignore_nothing
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso]: Maybe worth adding a TODO regarding order_by.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso [hunk 20]
```diff
@@ -142,3 +162,48 @@ resolve_aggregate table problem_builder aggregate_column =
...
+parse_aggregated_problems : Any->Vector
+parse_aggregated_problems problems =
+    if problems.is_nothing then [] else
+        problems_array = problems.getProblems
+        parsed = Vector.new problems_array.length i->
+            p = problems_array.at i
+            if Java.is_instance p InvalidAggregation then Invalid_Aggregation p.getColumnName p.getRow p.getMessage else
+                if Java.is_instance p FloatingPointGrouping then Floating_Point_Grouping p.getColumnName p.getRow else
+                    if Java.is_instance p UnquotedDelimiter then Unquoted_Delimiter p.getColumnName p.getRow else
+                        "Unknown Error"
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Aggregate_Column_Helper.enso]:
```suggestion
                        Illegal_State_Error "Could not recognize a reported problem coming from the Java `org.enso.table` library."
```

I think we shouldn't report warnings as pure Text, much like with dataflow errors - it makes it hard to match specific types etc. If a generic one or an 'impossible state' I think `Illegal_State_Error` is a decent fallback. Also I'd add some detail so that we know what has happened - `Unknown error` is rather vague.

File: std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java (hunk 7 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: Still think a comment here would be good just to make the intention clear.

File: std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java (hunk 8 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: Any thoughts on this?

File: std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+  @Override
+  public Object aggregate(List<Integer> indexes) {
+    StringBuilder current = null;
+    for (int row: indexes) {
+      Object value = storage.getItemBoxed(row);
+      if (value == null || value instanceof String) {
+        String textValue = toQuotedString(value, quote, join);
+
+        if (quote.equals("") && textValue.contains(join)) {
+          this.addProblem(new UnquotedDelimiter(this.getName(), row, "Unquoted delimiter."));
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java:36]: I think it would be better to have a counter of these problems and only create actual classes at the very end in a 'summarize' step.

Then you could also easily aggregate them - for example instead of 100 `UnquotedDelimiter` instances, you could do `UnquotedDelimiter which_rows` in one instance. Much in the same way as we already do for input problems like out of bound indices where we emit one aggregated problem instead of a series of them.

Plus, probably  could be a bit more performant.

File: std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+      }
+    }
+
+    if (current == null) {
+      return null;
+    }
+
+    if (prefix != null) { current.insert(0, prefix); }
+    current.append(suffix);
+    return current.toString();
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java:58]: If there are no rows this will return `null`, right? Wasn't it supposed to always return `prefix+suffix`?

File: std-bits/table/src/main/java/org/enso/table/aggregations/CountEmpty.java [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+import org.enso.table.data.column.storage.Storage;
+import org.enso.table.data.table.Column;
+import org.enso.table.data.table.problems.InvalidAggregation;
+
+import java.util.List;
+
+/***
+ * Aggregate Column counting the number of (not-)null entries in a group.
+ */
+public class CountEmpty extends AggregateColumn {
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/CountEmpty.java]: I'd add a short note specifying the behavior depending on `isEmpty` value. Without it, the current name is a bit confusing, at least I was confused when reading it.

File: std-bits/table/src/main/java/org/enso/table/aggregations/Last.java [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+    Object current = null;
+    for (int row: indexes) {
+      Object value = storage.getItemBoxed(row);
+      if (!ignoreNothing || value != null) {
+        current = value;
+      }
+    }
+    return current;
+  }
+}
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/Last.java]: no newline at EOF

File: std-bits/table/src/main/java/org/enso/table/aggregations/Last.java [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+  public Object aggregate(List<Integer> indexes) {
+    Object current = null;
+    for (int row: indexes) {
+      Object value = storage.getItemBoxed(row);
+      if (!ignoreNothing || value != null) {
+        current = value;
+      }
+    }
+    return current;
+  }
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/Last.java]: Probably not the biggest issue, but any reason why we need to traverse the whole list? If we start traversal from back we could do an early stop at first non-null value. Feels like it could be much faster.

File: std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  @Override
+  public Object aggregate(List<Integer> indexes) {
+    Object current = null;
+    for (int row: indexes) {
+      Object value = storage.getItemBoxed(row);
+      if (value != null) {
+        try {
+          if (current == null || Compare(current, value) == minOrMax) {
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java]: Would be good to explain what values `minOrMax` takes, as it is not immediately clear.

Also, I think Java does only require `compareTo` (which we delegate to) to return some positive or negative value - but it does not specify them to be `-1` and `1` and I'm pretty sure there are cases when this can indeed be a different value - in such cases this computation will be incorrect. I think it should be changed to check for the sign.

File: std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+        }
+      }
+    }
+    return current;
+  }
+
+  private static int Compare(Object current, Object value) {
+    if (current instanceof String && value instanceof String) {
+      return ((String)value).compareTo((String)current);
+    }
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java:50]: This will not be aligned with our String normalization.

I.e. for strings `'ś'` and `'s\u{301}'` it will treat them as unequal even though in-Enso they are equal. I think we should strive to be consistent.
Not saying this must necessarily be done in this PR, but I think we should at least open a ticket to fix this.

Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java:50]: I'm wondering if we could just make `std-table` depend on `std-base` so we can easily re-use all the Unicode handling helpers that we define there. I think this should be the right approach here (as we don't want to re-implement them, as this will easily get out of sync).

File: std-bits/table/src/main/java/org/enso/table/aggregations/ShortestOrLongest.java [hunk 28]
```diff
@@ -0,0 +1,58 @@
...
+    BreakIterator iter = BreakIterator.getCharacterInstance();
+    iter.setText(text);
+
+    int count = 0;
+    for (int end = iter.next(); end != BreakIterator.DONE; end = iter.next()) {
+      count++;
+    }
+
+    return count;
+  }
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/ShortestOrLongest.java:57]: Would be ideal to reuse the functions from `std-base`.

File: distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso (hunk 16 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Table/0.0.0-dev/src/Error.enso]: Improved text but yes I think display and details on warnings will be a good thing when we get them to UI.

File: std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java (hunk 22 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java:58]: no if no rows should be null.

File: std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java (hunk 27 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java:50]: will add a ticket

Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/MinOrMax.java:50]: https://www.pivotaltracker.com/story/show/181754646

File: std-bits/table/src/main/java/org/enso/table/aggregations/ShortestOrLongest.java (hunk 28 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/ShortestOrLongest.java:57]: https://www.pivotaltracker.com/story/show/181754646

File: distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso [hunk 29]
```diff
@@ -487,49 +486,47 @@ type Table
...
+         The following problems can occur:
+         - If a column name is not in the input table, a `Missing_Input_Columns`.
+         - If a column index is out of range, a `Column_Indexes_Out_Of_Range`.
+         - If there are no valid columns in the output table, a `No_Output_Columns`.
+         - If there are invalid column names in the output table, a `Invalid_Output_Column_Names`.
+         - If there are duplicate column names in the output table, a `Duplicate_Output_Column_Names`.
+         - If grouping on or computing the `Mode` on a floating point number, a `Floating_Point_Grouping`.
+         - If an aggregation fails, an `Invalid_Aggregation_Method`.
+         - If when concatenating values there is an quoted delimited, an `Unquoted_Delimiter`
+         - If there are more than 10 issues with a single column, an `Additional_Warnings`.
```
Comment (radeusgd) [inline distribution/lib/Standard/Table/0.0.0-dev/src/Data/Table.enso:508]: If each column can issue this thing, then I think it should contain the origin column as payload.

But I still think we should aggregate these issues in the same way as we do for inputs.

File: std-bits/table/src/main/java/org/enso/table/aggregations/CountDistinct.java [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.Set;
+
+/***
+ * Aggregate Column counting the number of distinct items in a group.
+ */
+public class CountDistinct extends Aggregator {
+  private final Storage[] storage;
+  private final boolean ignoreEmpty;
+
+  public CountDistinct(String name, Column[] columns, boolean ignoreEmpty) {
```
Comment (radeusgd) [inline std-bits/table/src/main/java/org/enso/table/aggregations/CountDistinct.java]: I think `ignoreEmpty` should be documented to indicate what does it mean.

Also shouldn't it be `ignoreAllNulls`? Empty has a different meaning in these functions than what you have here which is a potential source of confusion

File: test/Benchmarks/src/Table/Aggregate.enso [hunk 31]
```diff
@@ -30,20 +30,26 @@ main =
...
-    Bench.measure (table.aggregate [Group_By "Index", Median "Value"]) "Median grouped" here.iter_size here.num_iterations
-    Bench.measure (table.aggregate [Group_By "Index", Mode "Index"]) "Mode grouped" here.iter_size here.num_iterations
-
-    Bench.measure (table.aggregate [Group_By "Index", Group_By "Flag", Count Nothing]) "Count 2 level groups" here.iter_size here.num_iterations
-    Bench.measure (table.aggregate [Group_By "Index", Group_By "Flag", Count_Distinct "Code"]) "Count Distinct 2 level groups" here.iter_size here.num_iterations
-    Bench.measure (table.aggregate [Group_By "Index", Group_By "Flag", Standard_Deviation "Value"]) "StDev 2 level groups" here.iter_size here.num_iterations
-    Bench.measure (table.aggregate [Group_By "Index", Group_By "Flag", Median "Value"]) "Median 2 level groups" here.iter_size here.num_iterations
-    Bench.measure (table.aggregate [Group_By "Index", Group_By "Flag", Mode "Index"]) "Mode 2 level groups" here.iter_size here.num_iterations
+    Bench.measure (table.aggregate_2 [Count Nothing]) "Count table" here.iter_size here.num_iterations
```
Comment (radeusgd) [inline test/Benchmarks/src/Table/Aggregate.enso]:
```suggestion
    Bench.measure (table.aggregate [Count Nothing]) "Count table" here.iter_size here.num_iterations
```

wasn't the `_2` suffix just removed?

**Review by radeusgd — APPROVED** [review]
Looks good overall, although as noted some stuff like comparing performance or ensuring handling of Unicode strings need to be done at some point - but as agreed - will be more efficient to do these separately.

(still would appreciate addressing the earlier comments in one way or another)

File: std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java (hunk 22 above)
Comment (radeusgd) [reply std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java:58]: Oh that is news to me.

I think our `Aggregate_Column.enso` definitions need some more detailed documentation, because I think the expected behaviour in these kinds of edge cases should be documented, so that both users and implementators of alternative backends can know what to expect.

File: std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java (hunk 8 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: BigIntegers do not work in the table at the moment:
image

I will add a ticket to look at BigInteger, Date and BigDecimal support in Tables but this may be dependent on work by the engine team as well.

Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/AggregateColumn.java]: https://www.pivotaltracker.com/story/show/181755990

File: std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java (hunk 22 above)
Comment (jdunkerley) [reply std-bits/table/src/main/java/org/enso/table/aggregations/Concatenate.java:58]: Added extra comments to the columns.

**Review by 4e6 — APPROVED** [review]
